apps/studio/src/server/modules/auth/email/email.router.ts

```diff
@@ -95,34 +117,32 @@ export const emailSessionRouter = router({
         throw e
       }
 
-      const emailName = email.split("@")[0] ?? "unknown"
-
-      // Not using Prisma's `upsert` because Prisma's unique constraint with nullable fields
-      // like `deletedAt` causes type issues. Prisma expects `deletedAt` to be `string|Date`
-      // even when `null` is valid in the database schema.
-      const existingUser = await ctx.prisma.user.findFirst({
-        where: {
+      return db.transaction().execute(async (tx) => {
+        const user = await upsertUser({
+          tx,
           email,
-          deletedAt: null,
-        },
-      })
-      const user = existingUser
-        ? await ctx.prisma.user.update({
-            where: { id: existingUser.id },
-            data: { lastLoginAt: new Date() },
-            select: defaultMeSelect,
-          })
-        : await ctx.prisma.user.create({
-            data: {
-              email,
-              phone: "", // TODO: add the phone in later, this is a wip
-              name: emailName,
+        })
+
+        await logAuthEvent(tx, {
+          by: user,
+          delta: {
+            before: {
+              token: oldVerificationToken,
             },
-            select: defaultMeSelect,
-          })
+            after: {
+              token: {
+                ...oldVerificationToken,
+                attempts: oldVerificationToken.attempts + 1,
+              },
```

Comment:
not quite indicative of what's in our db - we delete the token on successful login. you can check this in `verifyToken`; because our audit logs are meant to be as reflective of our "state of the world" (ie, our db), wdyt about setting `after: null`

Reply:
Yeah I think setting `after: null` makes sense. Updated in 6df2f491137cb7321046a905680739bb66f3f381.